### PR TITLE
MAGECLOUD-2903: Add PHP Version Validation

### PR DESCRIPTION
**Pull request by arhiopterecs**
### Description
Adds verification of the current PHP version.

### Fixed Issues (if relevant)
1. magento/ece-tools#MAGECLOUD-2903: Add PHP Version Validation
https://magento2.atlassian.net/browse/MAGECLOUD-2903
### Manual testing scenarios
https://jira.corp.magento.com/browse/MAGETWO-97152
### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/ece-tools?pullRequest=396) <br/>All committers have signed the CLA.

File: src/Config/Validator/Deploy/PhpVersion.php [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+use Magento\MagentoCloud\Config\ValidatorInterface;
+use Magento\MagentoCloud\Config\Validator;
+use Magento\MagentoCloud\Package\MagentoVersion;
+use Magento\MagentoCloud\Package\Manager as PackageManager;
+use Magento\MagentoCloud\Config\Validator\Deploy\PhpVersion\ConstraintFactory;
+use Psr\Log\LoggerInterface;
+use Composer\Semver\Semver;
+
+/**
+ * Validates PHP version
```
Comment (NadiyaS) [inline src/Config/Validator/Deploy/PhpVersion.php]: Please provide more detailed description. What rules of validation

File: src/Config/Validator/Deploy/PhpVersion.php [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                );
+            }
+        } catch (\Exception $e) {
+            $this->logger->warning('Can\'t validate version of PHP: ' . $e->getMessage());
+        }
+        return $this->resultFactory->success();
+    }
+
+    /**
+     * Returns the latest PHP constraint
```
Comment (NadiyaS) [inline src/Config/Validator/Deploy/PhpVersion.php]: probably in this case better to say "highest", not "latest"

File: src/Config/Validator/Deploy/PhpVersion.php [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+     *
+     * @return ConstraintInterface
+     * @throws \Exception
+     */
+    private function getRecommendedPhpConstraint(): ConstraintInterface
```
Comment (NadiyaS) [inline src/Config/Validator/Deploy/PhpVersion.php]: "Recommended constraint" sounds strange. Please rename the method

File: src/Config/Validator/Deploy/PhpVersion/ConstraintFactory.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    /**
+     * @param string $operator
+     * @param string $version
+     * @return Constraint
+     */
+    public function constraint(string $operator, string $version): Constraint
+    {
+        return $this->container->create(Constraint::class, [$operator, $version]);
```
Comment (oshmyheliuk) [inline src/Config/Validator/Deploy/PhpVersion/ConstraintFactory.php]: This causing an exception:
`Unresolvable dependency resolving [Parameter #0 [ <required> $operator ]] in class Composer\Semver\Constraint\Constraint`
Tried locally and on cloud environment.
Magento 2.1.16

File: src/Config/Validator/Deploy/PhpVersion.php (hunk 1 above)
Comment (arhiopterecs) [reply src/Config/Validator/Deploy/PhpVersion.php]: Done

File: src/Config/Validator/Deploy/PhpVersion/ConstraintFactory.php (hunk 4 above)
Comment (arhiopterecs) [reply src/Config/Validator/Deploy/PhpVersion/ConstraintFactory.php]: Fixed

**Comment by andriyShevtsov** [comment]
@arhiopterecs could you please clarify why  validation occur on deploy phase ? I assume that we can do it during build, isn't it?

File: src/Config/Validator/Deploy/PhpVersion.php [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    /**
+     * Validates PHP version
+     *
+     * @return Validator\ResultInterface
+     */
+    public function validate(): Validator\ResultInterface
+    {
+        try {
```
Comment (shiftedreality) [inline src/Config/Validator/Deploy/PhpVersion.php:75]: This all can be replaced with next logic:
```php
$constraint = $this->composer->getLocker()->getLockedRepository()->findPackage('magento/magento2-base', '*')->getRequires()['php']->getConstraint()->getPrettyString();
$constraint = "LAST ITEM AFTER PIPE HERE";

die(var_dump(Semver::satisfies($currentPhpVersion, $constraint)));
```

**Review by shiftedreality — CHANGES_REQUESTED** [review]

File: src/Config/Validator/Deploy/PhpVersion.php (hunk 5 above)
Comment (michaelcasey316) [reply src/Config/Validator/Deploy/PhpVersion.php:75]: If we end up choosing that replacement, the first line needs line breaks for readability:
```php
$constraint = $this->composer
    ->getLocker()
    ->getLockedRepository()
    ->findPackage('magento/magento2-base', '*')
    ->getRequires()['php']
    ->getConstraint()
    ->getPrettyString();
$constraint = "LAST ITEM AFTER PIPE HERE";

die(var_dump(Semver::satisfies($currentPhpVersion, $constraint)));
```

Comment (shiftedreality) [reply src/Config/Validator/Deploy/PhpVersion.php:75]: There is wrapper exists

```
$requirePackages = $this->packageManager->get('magento/magento2-base')->getRequires();
$phpConstraint = $requirePackages['php']->getConstraint()->getPrettyString();
```

**Review by shiftedreality — APPROVED** [review]

**Comment by andriyShevtsov** [comment]
QA approved

File: src/Config/Validator/Deploy/PhpVersion.php [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            $recommendedPhpConstraint = $this->getRecommendedPhpConstraint();
+            $currentPhpConstraint = $this->getCurrentPhpConstraint();
+
+            if (!$recommendedPhpConstraint->matches($currentPhpConstraint)) {
+                return $this->resultFactory->error(
+                    sprintf(
+                        'For Magento %s recommended PHP version satisfying the constraint %s. '
```
Comment (meker12) [inline src/Config/Validator/Deploy/PhpVersion.php:82]: 'Magento %s requires PHP version %s. '

File: src/Config/Validator/Deploy/PhpVersion.php [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                        . 'Currently installed PHP version %s',
```
Comment (meker12) [inline src/Config/Validator/Deploy/PhpVersion.php:83]: 'The currently installed PHP version ( %s) does not meet that requirement',

File: src/Config/Validator/Deploy/PhpVersion.php [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+                        $this->magentoVersion->getVersion(),
+                        $recommendedPhpConstraint->getPrettyString(),
+                        $currentPhpConstraint->getPrettyString()
+                    ),
+                    "Change the version of PHP to the version that satisfies the restriction conditions.\n"
```
Comment (meker12) [inline src/Config/Validator/Deploy/PhpVersion.php:88]: "Change the PHP configuration in the magento.app.yaml file to specify a version that meets the requirement,\n"

File: src/Config/Validator/Deploy/PhpVersion.php [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+                    . 'Change the PHP version in the .magento.app.yaml file or create a support ticket'
```
Comment (meker12) [inline src/Config/Validator/Deploy/PhpVersion.php:89]: . 'or submit a support ticket requesting a PHP version change.'

File: src/Config/Validator/Deploy/PhpVersion.php [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+                );
+            }
+        } catch (\Exception $e) {
+            $this->logger->warning('Can\'t validate version of PHP: ' . $e->getMessage());
```
Comment (meker12) [inline src/Config/Validator/Deploy/PhpVersion.php:93]: $this->logger->warning('Cannot validate PHP version: ' . $e->getMessage());

Possible clarification: 
Does this  exception mean "It is not possible to validate the PHP version", or "the installed PHP version does not meet Magento %s configuration requirements" ?

**Review by meker12 — COMMENTED** [review]
Added some suggestions for clarifying text in validation and exception messages.

**Pull request merged**